cdsapi/api.py

```diff
@@ -101,60 +121,58 @@ def _download(self, url, size, target):
-            r = self.robust(self.session.get)(
-                url,
-                stream=True,
-                verify=self.verify,
-                headers=headers,
-                timeout=self.timeout,
-            )
-            try:
-                r.raise_for_status()
-
-                with tqdm(
-                    total=size,
-                    unit_scale=True,
-                    unit_divisor=1024,
-                    unit="B",
-                    disable=not self.progress,
-                    leave=False,
-                ) as pbar:
-                    pbar.update(total)
-                    with open(target, mode) as f:
+        with _open_target(target) as f:
+            while tries < self.retry_max:
+                r = self.robust(self.session.get)(
+                    url,
+                    stream=True,
+                    verify=self.verify,
+                    headers=headers,
+                    timeout=self.timeout,
+                )
+                try:
+                    r.raise_for_status()
+
+                    with tqdm(
+                        total=size,
+                        unit_scale=True,
+                        unit_divisor=1024,
+                        unit="B",
+                        disable=not self.progress,
+                        leave=False,
+                    ) as pbar:
+                        pbar.update(total)
                         for chunk in r.iter_content(chunk_size=1024):
                             if chunk:
                                 f.write(chunk)
                                 total += len(chunk)
                                 pbar.update(len(chunk))
 
-            except requests.exceptions.ConnectionError as e:
-                self.error("Download interupted: %s" % (e,))
-            finally:
-                r.close()
+                except requests.exceptions.ConnectionError as e:
+                    self.error("Download interupted: %s" % (e,))
+                finally:
+                    r.close()
 
-            if total >= size:
-                break
+                if total >= size:
+                    break
 
-            self.error(
-                "Download incomplete, downloaded %s byte(s) out of %s" % (total, size)
-            )
-            self.warning("Sleeping %s seconds" % (sleep,))
-            time.sleep(sleep)
-            mode = "ab"
-            total = os.path.getsize(target)
-            sleep *= 1.5
-            if sleep > self.sleep_max:
-                sleep = self.sleep_max
-            headers = {"Range": "bytes=%d-" % total}
-            tries += 1
-            self.warning("Resuming download at byte %s" % (total,))
+                self.error(
+                    "Download incomplete, downloaded %s byte(s) out of %s" % (total, size)
+                )
+                self.warning("Sleeping %s seconds" % (sleep,))
+                time.sleep(sleep)
+                total = f.tell()
```

Comment:
we may check the position of the write pointer now to find the total amount of bytes written.